src/tools/rbd_mirror/image_map/types.h

```diff
@@ -0,0 +1,161 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+
+#include <set>
+#include <map>
+#include <string>
+#include <memory>
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+enum MapShuffleType {
+  SHUFFLE_TYPE_INSTANCES_ADDED = 0,   // rebalance due to addition of instances
+  SHUFFLE_TYPE_INSTANCES_REMOVED,     // rebalance due to removal of instances
+  SHUFFLE_TYPE_IMAGES_ADDED_REMOVED,  // rebalance due to adding/removing images
+};
+
+struct MappedImage {
+  std::string id;
+  std::string global_id;
+  std::string instance_id;
+
+  MappedImage(const std::string &global_id)
+    : global_id(global_id) {
+  }
+
+  MappedImage(const std::string &id,
+              const std::string &global_id)
+    : id(id), global_id(global_id) {
+  }
+
+  MappedImage(
+    const std::string &id,
+    const std::string &global_id,
+    const std::string &instance_id)
+    : id(id),
+      global_id(global_id),
+      instance_id(instance_id) {
+  }
+
+  inline bool operator==(const MappedImage &rhs) const {
+    return (global_id == rhs.global_id && id == rhs.id);
+  }
+
+  inline bool operator<(const MappedImage &rhs) const {
+    return global_id < rhs.global_id;
+  }
+};
+
+struct RemappedImage {
```

Comment:
Why not just track this all within the same data structure above? It would have a state of associating, unassociating, associated, and rebalancing. Probably should also track "next instance id" for the case of rebalancing so that your policy can re-execute while a rebalance is in-progress and make decisions based on the expected state, not current state. Also should add a "last mapped" timestamp to ensure you don't ping-pong images back-and-force (i.e. a given image that is already mapped cannot be re-mapped for X minutes -- config option).